pkg/kubelet/images/image_pull_manager.go

```diff
@@ -0,0 +1,663 @@
+	} else {
+		if newCreds.NodePodsAccessible {
+			selectedCreds.NodePodsAccessible = true
+			selectedCreds.KubernetesSecrets = nil
+		} else {
+			selectedCreds.KubernetesSecrets = mergePullSecrets(selectedCreds.KubernetesSecrets, newCreds.KubernetesSecrets)
+		}
+		ret.CredentialMapping[imageNoTagDigest] = selectedCreds
+	}
+
+	updated := !reflect.DeepEqual(orig.CredentialMapping, ret.CredentialMapping)
+	if updated {
+		ret.LastUpdatedTime = metav1.Time{Time: time.Now()}
+	}
+	return *ret, updated
+}
+
+// mergePullSecrets merges two slices of ImagePullSecret object into one while
+// keeping the objects unique per `Namespace, Name, UID` key.
+//
+// In case an object from the `new` slice has the same `Namespace, Name, UID` combination
+// as an object from `orig`, the result will use the CredentialHash value of the
+// object from `new`.
+//
+// The returned slice is sorted by Namespace, Name and UID (in this order).
+func mergePullSecrets(orig, new []imagemanagerv1alpha1.ImagePullSecret) []imagemanagerv1alpha1.ImagePullSecret {
+	credSet := make(map[kubSecretCoordinates]string)
+	for _, secret := range orig {
+		credSet[kubSecretCoordinates{
+			UID:       secret.UID,
+			Namespace: secret.Namespace,
+			Name:      secret.Name,
+		}] = secret.CredentialHash
+	}
+
+	for _, s := range new {
+		credSet[kubSecretCoordinates{UID: s.UID, Namespace: s.Namespace, Name: s.Name}] = s.CredentialHash
+	}
+
+	ret := make([]imagemanagerv1alpha1.ImagePullSecret, 0, len(credSet))
+	for coords, hash := range credSet {
+		ret = append(ret, imagemanagerv1alpha1.ImagePullSecret{
+			UID:            coords.UID,
+			Namespace:      coords.Namespace,
+			Name:           coords.Name,
+			CredentialHash: hash,
+		})
+	}
+	// we don't need to use the stable version because secret coordinates used for ordering are unique in the set
+	slices.SortFunc(ret, imagePullSecretLess)
+
+	return ret
+}
+
+func cacheFilename(image string) string {
+	return fmt.Sprintf("sha256-%x", sha256.Sum256([]byte(image)))
+}
+
+// imagePullSecretLess is a helper function to define ordering in a slice of
+// ImagePullSecret objects.
+func imagePullSecretLess(a, b imagemanagerv1alpha1.ImagePullSecret) int {
+	if a.Namespace < b.Namespace {
+		return -1
+	} else if a.Namespace > b.Namespace {
+		return 1
+	}
+
+	if a.Name < b.Name {
+		return -1
+	} else if a.Name > b.Name {
+		return 1
+	}
+
+	if a.UID < b.UID {
+		return -1
+	}
+	return 1
+}
+
+// TODO: should we salt the hashes to prevent on-disk introspection for identical secrets?
+func hashImageCredentials(creds credentialprovider.AuthConfig) (string, error) {
+	credBytes, err := json.Marshal(creds)
+	if err != nil {
+		return "", err
+	}
+
+	hash := sha256.New()
+	hash.Write([]byte(credBytes))
+	// hash.Write([]byte(salt))
+	return fmt.Sprintf("%x", hash.Sum(nil)), nil
+}
+
+// writeFile writes `content` to the file with name `filename` in directory `dir`
+func writeFile(dir, filename string, content []byte) error {
```

Comment:
we have to do atomic replaces to avoid partially written data in files... that means write to a tmp file, then rename over the destination file